### ISSUE TITLE
memory leaks?

### ISSUE THREAD

**Issue by jwkvam** [cvxgrp/cvxpy#166]
I'm trying to hunt down some memory leaks in my program and it seems like there may be some with cvxpy or it's underlying solvers, I'm guessing the later is more likely. I basically copied the sample problem from the front page and repeatedly ran it and sampled the memory usage of the process and plotted the results for different solvers. For the ECOS and SCS solvers the memory has a very clear linear increase. CVXOPT seems fairly stable. Anyone encounter this before? Should I dig deeper?

```python
#!/usr/bin/env python -u
# encoding: utf-8

import numpy as np
from scipy import interpolate, ndimage
import gc
import os
import psutil
from cvxpy import *
import numpy

import objgraph

def sample_mem():
    gc.collect()
    process = psutil.Process(os.getpid())
    return process.get_memory_info()[0] / float(2 ** 20)

def cvx_solve():
    m = 128
    n = 10
    np.random.seed(1)
    A = np.random.randn(m, n)
    b = np.random.randn(m)

    # Construct the problem.
    x = Variable(n)
    objective = Minimize(norm(A*x - b))
    constraints = [0 <= x, x <= 1]
    prob = Problem(objective, constraints)
    prob.solve('SCS')
    # prob.solve('ECOS')
    # prob.solve('CVXOPT')

if __name__=='__main__':
    for _ in range(10000):
        cvx_solve()
        print(sample_mem())
```

![image](https://cloud.githubusercontent.com/assets/86304/6423971/f9255f20-bea3-11e4-9ded-8806c7e297ca.png)
![image](https://cloud.githubusercontent.com/assets/86304/6423973/fdd875de-bea3-11e4-832c-575b689f96d1.png)
![image](https://cloud.githubusercontent.com/assets/86304/6423974/003bbde0-bea4-11e4-9234-0556c7a630e1.png)

**Comment by echu** [comment]
IIRC, when the ECOS module returns the solution, it runs "cleanup" with a
flag telling it that the memory is going to be owned by "someone else". It
definitely does this in Matlab, but I'm not sure about Python.

If this is the source of the issue, I could run your code with a full
cleanup and see if the issue persists.

A simple fix would just be to copy the data into native (or numpy) data
structures.

Eric

**Comment by jwkvam** [comment]
Thanks @echu. I tried storing the result of `prob.solve()` and copying `soln = x.value.copy()` but get the same memory use.

**Comment by echu** [comment]
Hmm. I think there might still be an issue, but I set the ownership flag for the returned `x`, `s`, `y`, and `z` vectors to the numpy arrays (this is on the develop branch of [embotech/ecos-python](http://github.com/embotech/ecos-python)).

The memory usage still grows, but settles at 60 instead of ~100.

We've run valgrind on the C core of ECOS, and I believe we've eliminated any leaks there. This means any leaks come from our Python module. I'm not sure what the other possible source of leaks would be, so I'm open to suggestions there.

While it's possible CVXPY is causing these leaks, I don't think so, since we don't see the memory issue with CVXOPT. @SteveDiamond, you'll have to confirm / deny if you think CVXPY could be doing anything to keep memory around for ECOS/SCS but not for CVXOPT.

**Comment by bodono** [comment]
Thanks for letting us know about this @jwkvam. I implemented the change that @echu made in ECOS for SCS, which yields the memory to numpy properly, and I think it has helped (possibly fixed) the issue for SCS. Here is the same plot for SCS after the change:

![memusage](https://cloud.githubusercontent.com/assets/555071/6426361/3d482e44-bf4e-11e4-8e0e-1ba82bc29bf2.jpg)

Thanks @echu for altering me to the right way to handle the memory. By the looks of it CVXPY is probably doing what it should be doing, and there might be a small mem leakage in ECOS somewhere.

I've pushed this change to master. Let me know you find that there is still a memory problem in SCS after this.

**Comment by echu** [comment]
Great! I must've overlooked something while working on it then. :) I'll
take a second look at it today.

**Comment by echu** [comment]
See [embotech/ecos/issues/111](https://github.com/embotech/ecos/issues/111) for a partial resolution.

Same figure, but I'm still noticing a leak. I'll look into that a little bit later. It's a whole lot better now. :)

![figure_1](https://cloud.githubusercontent.com/assets/52689/6427745/d3dce616-bf3e-11e4-97b3-e389d45883ae.png)

If you want to be on the bleeding edge, you can clone this repo, switch the ecos submodule in this repo to the develop branch, and build from source.

**Comment by SteveDiamond** [comment]
Cvxpy is entirely in Python. The leak could be in a library cvxpy calls
(besides the solvers), but the fact that only ecos is leaking at this point
makes this unlikely.

**Comment by jwkvam** [comment]
Thanks guys for the quick turnaround, that's great.

**Comment by echu** [comment]
Okay. I found it. You happened to trigger a very obscure bug. :) I'm glad we caught it, though! (I can hear the ghost of formal methods folks over my shoulder....)

It's also resolved at embotech/ecos#111.
![figure_1](https://cloud.githubusercontent.com/assets/52689/6436163/c43e2e00-c05f-11e4-9cce-a3b198457da0.png)

**Comment by ihwang** [comment]
I seem to be seeing memory leaks for all solvers for the code snippet in the original post. I was told that this might now be a problem with [cvxcanon](https://github.com/cvxgrp/CVXcanon/issues/23), but haven't gotten much a response there, so I thought I'd repost to this issue thread. Is it possible that this bug has been re-triggered by other revisions since this issue was closed last year?
